# 단일표본 t검정

- 표본 데이터 로딩
- 표본 데이터가 정규성 따르는지 검정하기
- 단일표본 t검정 \
    - p-value < 0.05 : 귀무가설 기각, 대립가설 채택, 대립가설이 통계적으로 유의미 해석 가능
    - p-value >= 0.05 : 귀무가설 채택, 대립가설 기각

## 정규성 검정하기

In [1]:
float('150.27')

150.27

In [3]:
# data 불러오기
with open("./datas2/성인여성_키_데이터1.txt", "r") as f:
    data = f.read().split('\n')
    #print(list(map(float, data)))
    data = list(map(float, data))
    print(data) 

[163.94, 162.48, 164.29, 166.3, 162.26, 162.26, 166.43, 164.57, 161.72, 164.05, 161.73, 161.73, 163.36, 158.4, 158.83, 161.51, 160.47, 163.52, 160.71, 159.55, 166.17, 162.28, 162.96, 159.52, 161.55, 163.06, 160.15, 163.66, 161.42, 162.13]


## 데이터의 평균과 표준편차 구하기

In [4]:
from scipy import stats
import numpy as np

# 평균과 표준편차 계산
mean_value = np.mean(data)
std_dev = np.std(data, ddof=0)  # 모집단 표준편차(기본 설정)

mean_value, std_dev

(np.float64(162.36700000000002), np.float64(2.0354559685731335))

## 가설 설정
- 귀무가설: 표본 데이터의 평균은 163과 같다.
- 대립가설: 표본 데이터의 평균은 163과 다르다.

## 정규성 검정
- 수집한 표본 데이터가 정규분포를 따르는지 검증한다.

## 방법 1 : KS-test

In [6]:
# 정규성 검정, KS-test(Kolmogorov-Smirnov) 검정
# KS-test는 주어진 표본 데이터가 특정 이론적인 분포(예: 정규분포)를 따르는지 검정하는 방법

# 비교 대상이 되는 정규분포에 표본 데이터의 평균과 표준편차를 사용
ks_statistic, p_value = stats.kstest(
    data,
    lambda x: stats.norm.cdf(
        x,
        loc=np.mean(data),
        scale=np.std(data, ddof=0)
    )
)

ks_statistic, p_value

(np.float64(0.08754216181498631), np.float64(0.9601287796358491))

In [8]:
stats.shapiro(data) 

ShapiroResult(statistic=np.float64(0.9750912943880133), pvalue=np.float64(0.6854198354912403))

In [10]:
import pandas as pd
import numpy as np

In [ ]:
# encoding = euc-kr로 되어있는데 이걸 지금 utf-8로 바꿔서 읽어야 함
df1 = pd.read_csv("./datas2/반별_점수_type1.csv", encoding = "utf-8")
# df1 = pd.read_csv("./datas2/반별_점수_type1.csv")
df1.head()

,반,점수
0,A,73
1,A,69
2,A,71
3,A,71
4,A,73


In [23]:
# A반의 데이터만 추출
group_A = df1['점수'].loc[df1['반'] == 'A'].values
# B반의 데이터만 추출
group_B = df1['점수'].loc[df1['반'] == 'B'].values

print(group_A)
print(group_B)

[73 69 71 71 73 67 73 69 62 74 68 66 70 82 70 65 76 73 58 81]
[63 56 73 61 55 77 75 65 61 55]


In [19]:
import numpy as np

# A반의 평균과 표준 편차
mean_A = np.mean(group_A)
std_A = np.std(group_A, ddof=1)  # 표본 표준 편차 (ddof=1)

# B반의 평균과 표준 편차
mean_B = np.mean(group_B)
std_B = np.std(group_B, ddof=1)  # 표본 표준 편차 (ddof=1)

# 결과 출력
print(f"A반 평균: {mean_A:.2f}, 표준 편차: {std_A:.2f}")
print(f"B반 평균: {mean_B:.2f}, 표준 편차: {std_B:.2f}")

A반 평균: 70.55, 표준 편차: 5.68
B반 평균: 64.10, 표준 편차: 8.28


## 정규성 검정
- 정규성 검정 가설수립
    - 영가설 : 두 그룹의 표본 데이터는 정규분포를 따른다.
    - 대립가설 : 두 그룹의 표본 데이터는 정규성을 따르지 않는다.

In [ ]:
# group_A와 group_B의 평균과 표준편차를 사용하여 정규성 검정 수행
print(stats.kstest(
  group_A, 
  stats.norm.cdf, 
  args=(np.mean(group_A), np.std(group_A, ddof=0))))

print(stats.kstest(
  group_B, 
  stats.norm.cdf, 
  args=(np.mean(group_B), np.std(group_B, ddof=0))))

KstestResult(statistic=np.float64(0.1290433386337495), pvalue=np.float64(0.8515822805406548), statistic_location=np.float64(73.0), statistic_sign=np.int8(1))
KstestResult(statistic=np.float64(0.17142174723652215), pvalue=np.float64(0.8842512533111235), statistic_location=np.float64(73.0), statistic_sign=np.int8(-1))


### 해석

- p-value >= 0.05 : 모두 정규 분포를 따른다고 판단
- 독립표본 t검정 할 수 있음.

- 방법2 : Shapiro-Wilk 검정

In [24]:
stat, p = stats.shapiro(group_A) 
print(f"group_A의 stat: {stat}, p: {p}")
stat, p = stats.shapiro(group_B) 
print(f"group_B의 stat: {stat}, p: {p}")

group_A의 stat: 0.9696965805424343, p: 0.7484525953609655
group_B의 stat: 0.8888437047871043, p: 0.16456076105698642


## 독립 표본 t-검정수행

## 등분산성 검정
- 가설 수립
    - 영가설(귀무가설) : 두 그룹의 분산이 유사하다.
    - 대립가설 : 두 그룹은 분산이 유지하지 않다.

In [26]:
stats.levene(group_A, group_B) 

LeveneResult(statistic=np.float64(2.033067087400979), pvalue=np.float64(0.1649640862221014))